### ISSUE TITLE
Api name

### ISSUE THREAD

**Issue by mweststrate** [mweststrate/immer#24]
I am not too excited about `immer` as function name as it is undescriptive. What should the idiomatic name be?

Transform? Update? withMutations? Make?

**Comment by mweststrate** [comment]
Produce? Next? Mutate? Modify?

**Comment by Gregjarvez** [comment]
I was thinking about it too. Since it was designed primarily with redux in mind. i think the idiomatic name should contain redux-<something>

**Comment by AriaFallah** [comment]
@Gregjarvez this isn't coupled to redux though. It's for anything that you need to update immutably.

Regarding the actual name, I think "Create the next immutable state" from the README really captures the essence of what `immer` does. Thus, if verbosity wasn't an issue, then `createNextStateImmutably` would perfectly describe it. However, I do think that's a bit wordy so I maybe something like `create`, `createNext`, or `nextState` would work.

@mweststrate  Of your suggestions, I like most of them except update, mutate, or modify because those clash with the immutable nature of the library.

**Comment by benbraou** [comment]
I like`withMutations ` because it is familiar (`immutable.js`)
`next` and `produce` also convey the puropose of the library

**Comment by hex13** [comment]
I named similar function `transform` in my `transmutable` library.
https://github.com/hex13/enter-ghost/tree/master/packages/transmutable
I think it's kind of have sense - we're transforming the state by giving recipe of "how to transform"

On the other side `redux.*` would be horrible name - what if somebody would like to use this library out of Redux context? Coupling is bad, even in naming.

`next` would look good in context of Redux reducers, but it would look weird in anywhere else. 

I think update, mutate, or modify are also good.

I came up also with `run` name (something like controlled run - here a state, run some code in context of this state).

Or `commit` (because we "commit" mutations - like in Vuex).

two or more words would be probably to verbose for helper library.

**Comment by alexkrolick** [comment]
Some ideas:

- `finalize`
- `finalizeDraft`
- `applyDraftChanges`
- `applyDiffFromDraft`
- `applyDiff`
- `newStateFromDraft`

I like having `draft` in there because it makes it clear that the **return value** of the 2nd argument (`fn(draftStart)`) is what is important.

**Comment by mweststrate** [comment]
Or just 

`import ✏ from "immer" `

```javascript
function insertItem(array, action) {
    return ✏(array, draft => {
        draft.splice(action.index, 0, action.item)
    })
}
```

:)

**Comment by mweststrate** [comment]
I'm thinking about `transform(state, recipe: draft => {})`, it is short, makes quite clear that it produces a next state from the current one, and makes it nicely explainable:

"_transform_ the current _state_ to produce the next state according to this _recipe_ that expresses how a _draft_ state should be changed"

**Comment by alexkrolick** [comment]
"Transform" implies mutations, to me.

`nextState(currentState, transform)`

🤔

**Comment by Gregjarvez** [comment]
if the draft eventually becomes the next state 
```
function insertItem(array, action) {
    return draft(array, nextState=> {
        draft.splice(action.index, 0, action.item)
    })
}
```

:)

**Comment by mweststrate** [comment]
I think @alexkrolick brings up a good point, transform might imply modification existing data for many. _produce_ probably makes it clearer something new is created

```
function insertItem(array, action) {
    return produce(array, draft => {
        draft.splice(action.index, 0, action.item)
    })
}
```

**Comment by AjaxSolutions** [comment]
Definitely not immer.

The Urban Dictionary defines immer as someone who IMs a lot.

**Comment by hex13** [comment]
two ideas: 
1. swap parameters (recipe, state) instead of other way around. It could be counterintuitive but it could help with partial application and reusing transformations.
2. allow for passing additional parameters to `recipe` function

```javascript
const immer = require('immer').default;
const {createStore} = require('redux');

// proposal: if immer api look this way (fn, state, ...args):
const transform = (fn, state, ...args) => immer(state, draft => fn(draft, ...args));

//-----------------------------------------------
const initialState = {
    animals: ['cat', 'dog']
};

const addAnimal = (state, action) => {
    state.animals.push(action.animal);
};

const reducer = transform.bind(null, (state,  action) => {
    switch (action.type) {
        case 'add':
            addAnimal(state, action);
            break;
    }
});

const store = createStore(reducer, initialState);
store.subscribe(() => {
    console.log("STATE", store.getState())
});

store.dispatch({type: 'add', animal: 'monkey'});
store.dispatch({type: 'add', animal: 'chicken'});

```

though this is one liner but having such things in library out-of-the-box could be nice. Or it could use curry instead of partial application


(related: integration with Redux https://github.com/mweststrate/immer/issues/31 
 https://github.com/mweststrate/immer/issues/26 )

**Comment by alexkrolick** [comment]
@hex13 file a new issue for that? A separate thread would be nice.

**Comment by mweststrate** [comment]
@hex13 indeed, deserves separate issue :) also see #31, quite related

**Comment by mweststrate** [comment]
`produce` it will be. It allows for the concept of _producers_ which nicely rhymes with _reducers_ :). Will soon update. Or if someone beats me to it that is fine as well ;-). (make sure to claim the issue)

**Comment by mweststrate** [comment]
Done
